File: tests/php/lib/test-suite-runner.php
<?php

require_once __DIR__ . '/assertion-failed.php';

class TestSuiteRunner {
    private $suiteName;
    private $testsPassed = 0;
    private $testsSkipped = 0;
    private $testsFailed = 0;
    private $quiet = false;

    public function __construct($suiteName) {
        $this->suiteName = $suiteName;
        $this->quiet = in_array('-q', $_SERVER['argv'], true);
    }

    public function test($description, $test) {
        $this->output("→ $description\n");
        try {
            call_user_func($test);
            ++$this->testsPassed;
            $this->output("✅ $description\n");
        } catch (AssertionFailed $e) {
            ++$this->testsFailed;
            fwrite(STDOUT, "❌ $description\n");
            fwrite(STDOUT, $e->getMessage() . "\n");
        }
    }

    public function skip($description, $reason, $test) {
        if (!is_callable($test)) {
            throw new InvalidArgumentException(
                "Skipped test must be callable: $description"
            );
        }

        $this->testsSkipped++;

        echo "⚠️ $description — $reason\n";
    }

    private function output($message) {
        if (!$this->quiet) {
            fwrite(STDOUT, $message);
        }
    }

    public function finish() {
        $total =
            $this->testsPassed +
            $this->testsSkipped +
            $this->testsFailed;

        $statisticsFile = getenv('TEST_STATISTICS_FILE');

        if ($statisticsFile !== false && $statisticsFile !== '') {
            $suite = getenv('TEST_SUITE_ID');

            $statistics = array(
                'suite' => $suite,
                'status' => $this->testsFailed > 0 ? 'failed' : 'passed',
                'tests' => array(
                    'run' => $total,
                    'skipped' => $this->testsSkipped,
                    'passed' => $this->testsPassed,
                    'failed' => $this->testsFailed
                )
            );

            $temporaryFile = $statisticsFile . '.tmp';

            file_put_contents(
                $temporaryFile,
                json_encode($statistics)
            );

            rename($temporaryFile, $statisticsFile);
        }

        fwrite(
            STDOUT,
            "$this->suiteName: $total tests run, " .
            "$this->testsPassed passed, " .
            "$this->testsSkipped skipped, " .
            "$this->testsFailed failed\n"
        );

        if ($this->testsFailed > 0) {
            exit(1);
        }
    }
}

File: tests/python/lib/test_suite_runner.py
import json
import os
import sys


class AssertionFailed(Exception):
    pass


class TestSuiteRunner:
    def __init__(self, suite_name):
        self.suite_name = suite_name
        self.tests_passed = 0
        self.tests_skipped = 0
        self.tests_failed = 0
        self.quiet = '-q' in sys.argv

    def test(self, description, operation=None):
        if operation is None:
            def register(test_operation):
                self.test(description, test_operation)
                return test_operation

            return register

        self._output('→ {}\n'.format(description))

        try:
            operation()
            self.tests_passed += 1
            self._output('✅ {}\n'.format(description))
        except AssertionFailed as error:
            self.tests_failed += 1
            sys.stdout.write('❌ {}\n'.format(description))
            sys.stdout.write('{}\n'.format(error))

    def skip(self, description, reason, operation=None):
        if operation is None:
            def register(test_operation):
                self.skip(
                    description,
                    reason,
                    test_operation,
                )
                return test_operation

            return register

        if not callable(operation):
            raise ValueError(
                'Skipped test must be callable: {}'.format(
                    description,
                )
            )
        
        self.tests_skipped += 1
        
        sys.stdout.write(
            '⚠️ {} — {}\n'.format(
                description,
                reason,
            )
        )

    def _output(self, message):
        if not self.quiet:
            sys.stdout.write(message)

    def finish(self):
        total = self.tests_passed + self.tests_skipped + self.tests_failed

        statistics_file = os.environ.get('TEST_STATISTICS_FILE')

        if statistics_file:
            statistics = {
                'suite': os.environ.get('TEST_SUITE_ID'),
                'status': 'failed' if self.tests_failed else 'passed',
                'tests': {
                    'run': total,
                    'skipped': self.tests_skipped,
                    'passed': self.tests_passed,
                    'failed': self.tests_failed,
                },
            }

            temporary_file = statistics_file + '.tmp'

            with open(temporary_file, 'w') as file:
                json.dump(
                    statistics,
                    file,
                    separators=(',', ':'),
                )

            os.replace(temporary_file, statistics_file)
        sys.stdout.write(
            '{}: {} tests run, {} passed, {} skipped, {} failed\n'.format(
                self.suite_name,
                total,
                self.tests_passed,
                self.tests_skipped,
                self.tests_failed,
            )
        )


        if self.tests_failed:
            sys.exit(1)
